text/2020-01-30-key-spaces.md

```diff
@@ -0,0 +1,181 @@
+# Title
+
+Key Spaces
+
+## Summary
+
+When multiple applications store data in TiKV, we need a clear pattern for avoiding conflicting use of the key space in TiKV.
+TiKV is a monolithic KV store, so each application must choose a key prefix.
+Key Spaces formalizes this process by registering a key space for an application in TiKV itself.
+
+When combined with authentication (not part of this proposal), this can secure an application's key space.
+
+Additionally, key spaces make it possible for TiKV to avoid conflicts during a data restore by changing the key prefix assigned to a key space.
+
+## Motivation
+
+The use case for this is running multiple applications against TiKV (not just a single TiDB).
+Note that we may be motivated to run multiple instances of a single application such as TiDB.
+
+* Avoiding even accidental conflicting use of key space in TiKV is currently ad-hoc and unreliable
+* Securing data in TiKV is difficult. A necessary first step is clearly defining the boundaries of data.
+
+Once key spaces are in place, all applications and TiKV itself should understand what data belongs to whom and how to avoid key space conflict.
+
+## Detailed design
+
+#### Terminology
+
+This proposal allows the total TiKV key space to be divided into individudal key spaces for applications.
+To refer to the total TiKV key space we will use the terminology "total TiKV key space".
+
+#### Scope
+
+This RFC will not propose "sub spaces", which would be a key space within a key space.
+I think this is a good idea, but I also think it can simply be proposed independently, although it may be better to discuss after the key spaces RFC is accepted so that it can leverage some of the key space implementation.
+
+This RFC will not discuss authentication.
+Authentication is necessary to provide any meaningful security for key spaces, but I can start a separate RFC for this.
+
+#### Client view
+
+When a client first connects to the TiKV system, it will register itself to get a key space.
+This happens in the metadata component (PD).
+All requests to the storage component (TiKV) will include the key space.
```

Comment:
We know which TiKV the request should be sent to depends on the whole key, the way you proposed will a big challenge for the current routing way.

Reply:
But a key space still resolves to a key range. So it should still be equivalent to sending an entire key prefix. What is the challenge?

Reply:
The request must know the whole key before sending to TiKV

Reply:
It can since the key space resolves to a prefix. But I assumed PD would tell the client which TiKV to send its requests to. Do we have some documentation on this aspect?

Reply:
I hope https://github.com/tikv/client-go/blob/master/locate/region_cache.go can show you some details

Reply:
I still don't understand the issue here since a key space is equivalent to a key prefix. Perhaps we need to develop this proposal to a greater level of detail.

Reply:
In short, client can't request TiKV `Put(keyspace, key, value)` directly, because you don't know which TiKV you should request, instead, you should handle the keyspace logic(maybe need to request PD for some meta-information, this can be handled in tikv-client) at first, and then you know which TiKV you should send the `Put(encoded key, value)` request to.

Reply:
Currently the client knows where to send the key because PD tells it? It seems that the client just needs to ask PD what TiKV a (keyspace, key) should be sent to instead?

It would be nice for the client to not have to know the actual prefix. But I think it may also possible for the client to fully resolve the key space to a prefix. When adding on a feature such as auth, TiKV would need to recognize the prefix as belonging to a key space.

Reply:
This RFC is changed now so that TiKV has no knowledge of key spaces. I created a new RFC for TiKV to assist with sending prefixes that can be used for key spaces: #56